File: statistical_outlier_filter/statistical_outlier_filter.py
#!/usr/bin/env python2
import pcl
import pcl.pcl_visualization
import numpy as np

import rospy
from sensor_msgs.msg import PointCloud2
import sensor_msgs.point_cloud2 as pc2

file_path = '/home/benlee/Desktop/pc_practice/pcd_files/data'
def read_pcd():
    print("read pcd")
    #profile_np = np.load(file_path + '/table_scene_lms400.npy')
    profile_pcd = pcl.load(file_path + '/table_scene_lms400.pcd')
    in_statistical_outlier_filter_pcd, out_statistical_outlier_filter_pcd = do_statistical_outlier_filter(profile_pcd, 50, 1.0)
    visualize(out_statistical_outlier_filter_pcd)

def visualize(pc_data):
    print("visualize start")
    #centred = pc_data - np.mean(pc_data, 0)
    #pc_centred = pcl.PointCloud()
    #pc_centred.from_array(centred)

    visual = pcl.pcl_visualization.CloudViewing()
    visual.ShowMonochromeCloud(pc_data,b'pc_data')
    v = True
    while v:
        v = not(visual.WasStopped())

def do_statistical_outlier_filter(pc_data, mean, std):
	print("do statistical outlier filter")
	fil = pc_data.make_statistical_outlier_filter()
	#gaussian setting
	fil.set_mean_k(mean)
	fil.set_std_dev_mul_thresh(std)

	outlier_filtered = fil.filter() 

	fil.set_negative(True)
	inliter_filtered = fil.filter()
	return inliter_filtered, outlier_filtered

if __name__=="__main__":
    read_pcd()

File: statistical_outlier_filter/statistical_outlier_filter.cpp
#include <pcl/visualization/cloud_viewer.h>
#include <pcl/filters/statistical_outlier_removal.h>
#include <iostream>
#include <pcl/io/io.h>
#include <pcl/io/pcd_io.h>
    
int user_data;

int main ()
{
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered (new pcl::PointCloud<pcl::PointXYZ>);

    pcl::io::loadPCDFile ("/home/benlee/Desktop/pc_practice/pcd_files/data/table_scene_lms400.pcd", *cloud);

    std::cerr << "Cloud before filtering: " << std::endl;
    std::cerr << *cloud << std::endl;
    // Create the filtering object
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
    sor.setInputCloud (cloud);
    sor.setMeanK (50);
    sor.setStddevMulThresh (1.0);
    sor.filter (*cloud_filtered);

    std::cerr << "Cloud after filtering: " << std::endl;
    std::cerr << *cloud_filtered << std::endl;

    pcl::visualization::CloudViewer viewer("Cloud Viewer");
    viewer.showCloud(cloud_filtered);
    while (!viewer.wasStopped ())
    {
        user_data++;
    }
    return 0;
}
